tls/s2n_kex.c

```diff
@@ -184,66 +213,112 @@ const struct s2n_kex s2n_hybrid_ecdhe_kem = {
+
+    return S2N_RESULT_OK;
 }
 
-int s2n_kex_server_key_recv_read_data(const struct s2n_kex *kex, struct s2n_connection *conn, struct s2n_blob *data_to_verify, struct s2n_kex_raw_server_data *raw_server_data)
+S2N_RESULT s2n_kex_server_key_recv_read_data(const struct s2n_kex *kex, struct s2n_connection *conn, struct s2n_blob *data_to_verify,
+        struct s2n_kex_raw_server_data *raw_server_data)
 {
-    notnull_check(kex);
-    notnull_check(kex->server_key_recv_read_data);
-    return kex->server_key_recv_read_data(conn, data_to_verify, raw_server_data);
+    ENSURE_REF(kex);
+    ENSURE_REF(kex->server_key_recv_read_data);
+    ENSURE_REF(conn);
+    ENSURE_REF(data_to_verify);
+
+    GUARD_AS_RESULT(kex->server_key_recv_read_data(conn, data_to_verify, raw_server_data));
+
+    return S2N_RESULT_OK;
 }
 
-int s2n_kex_server_key_send(const struct s2n_kex *kex, struct s2n_connection *conn, struct s2n_blob *data_to_sign)
+S2N_RESULT s2n_kex_server_key_send(const struct s2n_kex *kex, struct s2n_connection *conn, struct s2n_blob *data_to_sign)
 {
-    notnull_check(kex);
-    notnull_check(kex->server_key_send);
-    return kex->server_key_send(conn, data_to_sign);
+    ENSURE_REF(kex);
+    ENSURE_REF(kex->server_key_send);
+    ENSURE_REF(conn);
+    ENSURE_REF(data_to_sign);
+
+    GUARD_AS_RESULT(kex->server_key_send(conn, data_to_sign));
+
+    return S2N_RESULT_OK;
 }
 
-int s2n_kex_client_key_recv(const struct s2n_kex *kex, struct s2n_connection *conn, struct s2n_blob *shared_key)
+S2N_RESULT s2n_kex_client_key_recv(const struct s2n_kex *kex, struct s2n_connection *conn, struct s2n_blob *shared_key)
 {
-    notnull_check(kex);
-    notnull_check(kex->client_key_recv);
-    return kex->client_key_recv(conn, shared_key);
+    ENSURE_REF(kex);
+    ENSURE_REF(kex->client_key_recv);
+    ENSURE_REF(conn);
+    ENSURE_REF(shared_key);
+
+    GUARD_AS_RESULT(kex->client_key_recv(conn, shared_key));
+
+    return S2N_RESULT_OK;
 }
 
-int s2n_kex_client_key_send(const struct s2n_kex *kex, struct s2n_connection *conn, struct s2n_blob *shared_key)
+S2N_RESULT s2n_kex_client_key_send(const struct s2n_kex *kex, struct s2n_connection *conn, struct s2n_blob *shared_key)
 {
-    notnull_check(kex);
-    notnull_check(kex->client_key_send);
-    return kex->client_key_send(conn, shared_key);
+    ENSURE_REF(kex);
+    ENSURE_REF(kex->client_key_send);
+    ENSURE_REF(conn);
+    ENSURE_REF(shared_key);
+
+    GUARD_AS_RESULT(kex->client_key_send(conn, shared_key));
+
+    return S2N_RESULT_OK;
 }
 
-int s2n_kex_tls_prf(const struct s2n_kex *kex, struct s2n_connection *conn, struct s2n_blob *premaster_secret)
+S2N_RESULT s2n_kex_tls_prf(const struct s2n_kex *kex, struct s2n_connection *conn, struct s2n_blob *premaster_secret)
 {
-    notnull_check(kex);
-    notnull_check(kex->prf);
-    return kex->prf(conn, premaster_secret);
+    ENSURE_REF(kex);
+    ENSURE_REF(kex->prf);
+    ENSURE_REF(conn);
+    ENSURE_REF(premaster_secret);
```

Comment:
You don't dereference `conn` or `premaster_secret`, so you don't strictly need to ensure they're not null. Is there a reason to ENSURE_REF them? (same question for a bunch of the other methods in this file)

Reply:
True, they're not strictly needed here. I generally err on the side of aggressively notnull_checking/ENSURE_REFing for added protection, as long as the arguments are expected to be not null (like should be the case here). My reasoning is that 1) if the function gets refactored in the future and those arguments are dereferenced, it can be easy to overlook adding in the notnull check when doing the refactor; 2) if a function further down the call stack happens to be missing a notnull check, we'll catch it here gracefully.

Reply:
That's fair, although if applied to every function it could lead to us unnecessarily making the same ENSURE_REF check over and over as arguments are passed down the call stack without being used. But maybe that's a small enough price to pay :)

Reply:
Yeah without compiler-assisted null checking it's kind of what we're left with... <insert plug for Rust here />